src/persist-client/src/operators/shard_source.rs

```diff
@@ -345,7 +348,12 @@ where
             SnapshotMode::Include => match read.snapshot(as_of.clone()).await {
                 Ok(parts) => parts,
                 Err(e) => {
-                    panic!("{name_owned}: {shard_id} cannot serve requested as_of {as_of:?}: {e:?}")
+                    error_handler(format!("{name_owned}: {shard_id} cannot serve requested as_of {as_of:?}: {e:?}"));
+
+                    // We cannot continue, and we cannot shut down. Otherwise
+                    // downstream operators might interpret our downgrading our
+                    // sink as a statement of progress.
+                    future::pending().await
```

Comment:
It strikes me that every call to the `error_handler` should be accompanied by a `future::pending().await`, and forgetting the second bit could cause pretty gnarly issues. Thoughts on wrapping this up behind an interface so it's easier to enforce that the two always go together?

```
struct ErrorHandler { callback: Box<dyn FnMut...> }
impl ErrorHandler { async fn report_and_stop(...) }
```

(That specific thing is probably the wrong way to do it, but you get the idea.)

Reply:
I pushed something like this and will hit merge. 👌